<a href="https://colab.research.google.com/github/eduardozurek/ia/blob/main/_007_aprendizaje_supervisado/_006_07_Clasificador_KNN_Binarizando_las_categ%C3%B3ricas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Ejemplo - Categoricas Binarizadas - KNN**


In [ ]:
import pandas as pd
#url = "https://web.stanford.edu/~hastie/ElemStatLearn/datasets/marketing.data"
datos = pd.read_csv("marketing.data",sep=' ',header=None)

In [ ]:
datos

array([[ 9.,  2.,  1., ...,  1.,  7., nan],
       [ 9.,  1.,  1., ...,  1.,  7.,  1.],
       [ 9.,  2.,  1., ...,  3.,  7.,  1.],
       ...,
       [ 1.,  2.,  5., ...,  1.,  7.,  1.],
       [ 4.,  1.,  1., ...,  3.,  7.,  1.],
       [ 6.,  1.,  5., ...,  3.,  5.,  1.]])

In [ ]:
import numpy as np
datos = np.asarray(datos)
print(datos.shape)
print(datos)

(8993, 14)
[[ 9.  2.  1. ...  1.  7. nan]
 [ 9.  1.  1. ...  1.  7.  1.]
 [ 9.  2.  1. ...  3.  7.  1.]
 ...
 [ 1.  2.  5. ...  1.  7.  1.]
 [ 4.  1.  1. ...  3.  7.  1.]
 [ 6.  1.  5. ...  3.  5.  1.]]


Como se puede ver, este conjunto de datos contiene varios "nan".

Lo primero que debemos hacer es remmover esos "nan" para evitar errores.

In [ ]:
contar_nan=np.isnan(datos)
filas_con_nan = contar_nan.sum(axis=1)
print(filas_con_nan.shape)
print(np.sum(filas_con_nan>0))
print(np.sum(filas_con_nan==0))

(8993,)
2117
6876


Esto me dice que hay, 2117 filas que contienen "nan"\
y 6876 fila que no contienen "nan".

Los metodos de clasificación estudiados aquí\
no tienen manera de lidiar con los "nan",\
entonces trabajaremos solo con las filas que\
no tienen "nan"

In [ ]:
datos = datos[filas_con_nan==0,:]
# y volvemos a revisar
contar_nan=np.isnan(datos)
filas_con_nan = contar_nan.sum(axis=1)
print(filas_con_nan.shape)
print(np.sum(filas_con_nan>0))
print(np.sum(filas_con_nan==0))

(6876,)
0
6876


In [ ]:
datos

array([[9., 1., 1., ..., 1., 7., 1.],
       [9., 2., 1., ..., 3., 7., 1.],
       [1., 2., 5., ..., 1., 7., 1.],
       ...,
       [1., 2., 5., ..., 1., 7., 1.],
       [4., 1., 1., ..., 3., 7., 1.],
       [6., 1., 5., ..., 3., 5., 1.]])

Ahora sí ...\
Procedemos binarizar las categorías de X y a entrenar y probar el clasificador

In [ ]:
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()
y = datos[:,0]
print(y)
X = datos[:,1:15]
print(X.shape)
xt = X[:,0]
lb.fit(xt)
X_binarizada = lb.transform(xt)
for i in range(1,13):
  xt = X[:,i]
  lb.fit(xt)
  X_binarizada = np.append(X_binarizada,lb.transform(xt),axis=1)
print(X_binarizada.shape)
X_binarizada = X_binarizada>0
X=X_binarizada

[9. 9. 1. ... 1. 4. 6.]
(6876, 13)
(6876, 74)


In [ ]:
X_binarizada

array([[False,  True, False, ...,  True, False, False],
       [ True,  True, False, ...,  True, False, False],
       [ True, False, False, ...,  True, False, False],
       ...,
       [ True, False, False, ...,  True, False, False],
       [False,  True, False, ...,  True, False, False],
       [False, False, False, ...,  True, False, False]])

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.33, random_state=17)

In [ ]:
from sklearn import neighbors
clfKNN = neighbors.KNeighborsClassifier(n_neighbors=101, # número de vecinos
                                        weights='uniform', # pesos asignados a los vecinos al momento
                                                           # de tomar la decisión
                                                           # 'uniform', 'distance'
                                        algorithm='brute', # optimiza la ejecución del clasificador
                                                           # ‘auto’, ‘ball_tree’, ‘kd_tree’, ‘brute’
                                        metric='jaccard') #jaccard
y_esperado = y_test
clfKNN.fit(X_train,y_train)
y_predicho = clfKNN.predict(X_test)
print(X_train)

[[ True  True False ...  True False False]
 [ True False  True ...  True False False]
 [ True False False ...  True False False]
 ...
 [False False False ...  True False False]
 [ True  True False ...  True False False]
 [ True False False ...  True False False]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_esperado,y_predicho))

0.33524229074889866


La siguiente celda de código no es correcta porque se hizo mal la binarización

In [ ]:
# Esto no se debe hacer
# hay que tener cuidado con la binarización
X1 = X
from sklearn.model_selection import train_test_split
X1_train, X1_test, y1_train, y1_test = \
    train_test_split(X1, y, test_size=0.33, random_state=17)
from sklearn import neighbors
clfKNN = neighbors.KNeighborsClassifier(n_neighbors=101, # número de vecinos
                                        weights='uniform', # pesos asignados a los vecinos al momento
                                                           # de tomar la decisión
                                                           # 'uniform', 'distance'
                                        algorithm='brute', # optimiza la ejecución del clasificador
                                                           # ‘auto’, ‘ball_tree’, ‘kd_tree’, ‘brute’
                                        metric='jaccard')
y1_esperado = y1_test
clfKNN.fit(X1_train,y1_train)
y1_predicho = clfKNN.predict(X1_test)
print(X1_train)
from sklearn.metrics import accuracy_score
print(accuracy_score(y1_esperado,y1_predicho))

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/pairwise.py:1875: DataConversionWarning: Data was converted to boolean for metric jaccard
  warnings.warn(msg, DataConversionWarning)


[[2. 1. 4. ... 1. 3. 1.]
 [2. 2. 3. ... 1. 5. 1.]
 [2. 3. 6. ... 3. 5. 1.]
 ...
 [1. 3. 6. ... 1. 7. 1.]
 [2. 1. 4. ... 2. 7. 1.]
 [2. 5. 1. ... 1. 2. 1.]]
0.17533039647577092


Como se puede ver, el sistema es capaz de clasificar\
correctamente menos del 31% de los datos de prueba

Pero que pasa si probamos el sistema solo con\
las dos categorías más bajas y las dos más altas

Veamos ...

In [ ]:
# con la siguiente línea vemos las diferentes categorías
print(np.unique(y))
indice = (y==1.0)|(y==2.0)|(y==8.0)|(y==9.0)
y1 = y[indice]
X1 = X[indice,:]
print(X1.shape)
print(y1.shape)
X_train, X_test, y_train, y_test = \
    train_test_split(X1, y1, test_size=0.33, random_state=17)
clfKNN.fit(X_train,y_train)
y_predicho = clfKNN.predict(X_test)
y_esperado = y_test
print(accuracy_score(y_esperado,y_predicho))

[1. 2. 3. 4. 5. 6. 7. 8. 9.]
(3596, 74)
(3596,)
0.5964616680707666


El clasificador mejoró!

Veamos solo con la primera y la última categoría ...

In [ ]:
# con la siguiente línea vemos las diferentes categorías
print(np.unique(y))
indice = (y==1.0)|(y==9.0)
y1 = y[indice]
X1 = X[indice,:]
print(X1.shape)
print(y1.shape)
X_train, X_test, y_train, y_test = \
    train_test_split(X1, y1, test_size=0.33, random_state=17)
clfKNN.fit(X_train,y_train)
y_predicho = clfKNN.predict(X_test)
y_esperado = y_test
print(accuracy_score(y_esperado,y_predicho))

[1. 2. 3. 4. 5. 6. 7. 8. 9.]
(1998, 74)
(1998,)
0.8848484848484849


Mejoró mucho más!

El clasificador diferencia muy bien\
entre los que tienen\
los ingresos más bajos y los más altos

In [ ]:
# con la siguiente línea vemos las diferentes categorías
print(np.unique(y))
indice = (y==1.0)|(y==5.0)|(y==9.0)
y1 = y[indice]
X1 = X[indice,:]
print(X1.shape)
print(y1.shape)
X_train, X_test, y_train, y_test = \
    train_test_split(X1, y1, test_size=0.33, random_state=17)
clfKNN.fit(X_train,y_train)
y_predicho = clfKNN.predict(X_test)
y_esperado = y_test
print(accuracy_score(y_esperado,y_predicho))

[1. 2. 3. 4. 5. 6. 7. 8. 9.]
(2525, 74)
(2525,)
0.7541966426858513


In [ ]:
# con la siguiente línea vemos las diferentes categorías
print(np.unique(y))
indice = (y==1.0)|(y==2.0)|(y==8.0)|(y==9.0)
y1 = y[indice]
y1[(y1==1.0)|(y1==2.0)]=0.0
y1[(y1==8.0)|(y1==9.0)]=10.0
X1 = X[indice,:]
print(X1.shape)
print(y1.shape)
X_train, X_test, y_train, y_test = \
    train_test_split(X1, y1, test_size=0.33, random_state=17)
clfKNN.fit(X_train,y_train)
y_predicho = clfKNN.predict(X_test)
y_esperado = y_test
print(accuracy_score(y_esperado,y_predicho))

[1. 2. 3. 4. 5. 6. 7. 8. 9.]
(3596, 74)
(3596,)
0.8609941027801179
